### PR TITLE
Unify Quantization APIs for add, pool and relu

### PR DESCRIPTION
**Pull request by supriyar**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#26335 Unify Quantization APIs for add, pool and relu**
* #26307 Changes to support int8 weight and fp32 bias in QNNPACK

Summary:
Use the backend engine flag to call QNNPACK for quantized ops.

Test Plan:
python test/test_quantized.py TestQNNPACKOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D17504331](https://our.internmc.facebook.com/intern/diff/D17504331)

### CODE REVIEW

File: aten/src/ATen/native/quantized/cpu/qadd.cpp
Comment (jerryzh168) [inline aten/src/ATen/native/quantized/cpu/qadd.cpp]: nit: error message? also if createStatus == success, is this assert always true?

File: aten/src/ATen/native/quantized/cpu/qadd.cpp
Comment (jerryzh168) [inline aten/src/ATen/native/quantized/cpu/qadd.cpp]: I think this one might not be INTERNAL ASSERT since user might have invalid input etc.?

Also looks like L131-L137 are general to all qnnpack ops? should we refactor this to separate function?

Comment (supriyar) [reply aten/src/ATen/native/quantized/cpu/qadd.cpp]: We check for inputs separately at the start. This is an internal assert from something going wrong inside qnnpack.
I don't think refactoring line 131-137 is necessary at this point, as eventually we will change all ops API for qnnpack to make it more functional

File: aten/src/ATen/native/quantized/cpu/qadd.cpp
Comment (supriyar) [reply aten/src/ATen/native/quantized/cpu/qadd.cpp]: Yeah, this is just an additional check. Can be removed.

File: aten/src/ATen/native/quantized/cpu/qpool.cpp
Comment (raghuramank100) [inline aten/src/ATen/native/quantized/cpu/qpool.cpp]: Do you need a check for input layout here? How do you know that the input is nhwc contiguous?

File: test/test_quantized.py
Comment (raghuramank100) [inline test/test_quantized.py]: Similar comments to conv/linear qnnpack tests:  We should ensure that all backends share the same tests and have consistent numerics. Can you merge this test with the existing relu test?

**Review by raghuramank100 — CHANGES_REQUESTED** [review]
I have comments mainly on test coverage and on unifying tests between FBGEMM and QNNPACK.

File: test/test_quantized.py
Comment (supriyar) [reply test/test_quantized.py]: Discussed offline, will merge in a separate PR.

File: test/test_quantized.py
Comment (supriyar) [reply test/test_quantized.py]: Will update the zero point. The input is quint8 since that is the only data type qnnpack currently supports.

File: aten/src/ATen/native/quantized/cpu/qpool.cpp
Comment (supriyar) [reply aten/src/ATen/native/quantized/cpu/qpool.cpp]: I'll make the change similar to https://github.com/pytorch/pytorch/pull/26242 once that lands for all qnnpack ops.

**Comment by t-vi** [comment]
It might well be me, but on raspbian (32bit armv7 hf) and `USE_PYTORCH_QNNPACK=1`, I get None from conv_prepack called like this:
```
r = torch.ones([2,2,2,2], dtype=torch.float)
scale, zero_point = 1, 2
q = torch.quantize_linear(r, scale, zero_point, torch.qint8)
torch.ops.quantized.conv_prepack(q, None, [1,1], [0,0], [1,1], 1)
```
I do get some torch.uint8 tensor when I do the same with a vanilla build on x86 (i.e. with FBGEMM).

**Comment by supriyar** [comment]
> It might well be me, but on raspbian (32bit armv7 hf) and `USE_PYTORCH_QNNPACK=1`, I get None from conv_prepack called like this:
> 
> ```
> r = torch.ones([2,2,2,2], dtype=torch.float)
> scale, zero_point = 1, 2
> q = torch.quantize_linear(r, scale, zero_point, torch.qint8)
> torch.ops.quantized.conv_prepack(q, None, [1,1], [0,0], [1,1], 1)
> ```
> 
> I do get some torch.uint8 tensor when I do the same with a vanilla build on x86 (i.e. with FBGEMM).

Did you try setting torch.backends.quantized.engine = torch.qnnpack? You can set the same on server side (x86) as well to confirm qnnpack works.

**Comment by t-vi** [comment]
Indeed, this fixes the example case and seems to return something in the quantized model. Thank you!

At some point, `conv_prepack` might raise a more descriptive exception except of just returning None... :slightly_smiling_face: 
Should the engine default to qnnpack when it available but fbgemm isn't?

File: aten/src/ATen/native/quantized/cpu/qpool.cpp
Comment (supriyar) [inline aten/src/ATen/native/quantized/cpu/qpool.cpp]: @dzhulgakov please review this. I changed it to permute input tensor to NHWC similar to your changes for qconv.

Comment (dzhulgakov) [reply aten/src/ATen/native/quantized/cpu/qpool.cpp]: Yep, it should be fine

**Review by dzhulgakov — COMMENTED** [review]
Looks good, I'll let Jerry to review in the details.

Would be really nice to unify the tests too and extend the benchmarks in benchmarks/op_benchmarks to go through both engines

**Review by jerryzh168 — APPROVED** [review]
LGTM

**Pull request closed without merging**
